File: Java/Data Structure/Linked List/program2.java
//Doubly linear linked list

import java.util.*;

class node
{
    public int data;
    public node next;
    public node prev;
}

class DoublyLLL
{
    public node head;
    public int Count;

    public DoublyLLL()
    {
        head = null;
        Count = 0;
    }

    public void InsertFirst(int no)
    {
        node newn = new node();

        newn.data = no;
        newn.next = null;
        newn.prev = null;

        if(head == null)
        {
            head = newn;
        }
        else
        {
            newn.next = head;
            head.prev = newn;
            head = newn;
        }
        Count++;
    }

    public void InsertLast(int no)
    {
        node newn = new node();

        newn.data = no;
        newn.next = null;
        newn.prev = null;

        if(head == null)
        {
            head = newn;
        }
        else
        {
            node temp = head;

            while(temp.next != null)
            {
                temp = temp.next;
            }
            temp.next = newn;
            newn.prev = temp;
        }
        Count++;
    }

    public void Display()
    {
        node temp = head;

        if(Count == 0)
        {
            System.out.println("Linked list is empty");
            return;
        }

        System.out.println("Elements of the linked list are: ");

        while(temp != null)
        {
            System.out.print("|"+ temp.data + "| -> ");
            temp = temp.next;
        }
        System.out.println("NULL");
    }

    public int CountNode()
    {
        return Count;
    }

    public void InsertAtPos(int no,int Pos)
    {
        if(Pos < 1 || Pos > Count+1)
        {
            System.out.println("Invalid Position");
            return;
        }

        if(Pos == 1)
        {
            InsertFirst(no);
        }
        else if(Pos == Count+1)
        {
            InsertLast(no);
        }
        else
        {
            node newn = new node();

            newn.data = no;
            newn.next = null;
            newn.prev = null;

            node temp = head;

            for(int i = 1;i < Pos-1; i++)
            {
                temp = temp.next;
            }
            newn.next = temp.next;
            temp.next.prev = newn;
            temp.next = newn;
            newn.prev = temp;

            Count++;
        }
    }

    public void DeleteFirst()
    {
        if(Count == 0)
        {
            return;
        }
        else if(Count == 1)
        {
            head = null;
        }
        else
        {
            head = head.next;
            head.prev = null;
        }

        Count--;
    }

    public void DeleteLast()
    {
        if(Count == 0)
        {
            return;
        }
        else if(Count == 1)
        {
            head = null;
        }
        else 
        {
            node temp = head;

            while(temp.next.next != null)
            {
                temp = temp.next;
            }
            temp.next = null;
        }

        Count--;
    }

    public void DeleteAtPos(int Pos)
    {
        if(Pos < 1 || Pos > Count)
        {
            System.out.println("Invalid Position");
            return;
        }

        if(Pos == 1)
        {
            DeleteFirst();
        }
        else if(Pos == Count)
        {
            DeleteLast();
        }
        else
        {
            node temp = head;

            for(int i = 1; i < Pos-1; i++)
            {
                temp = temp.next;
            }
            temp.next = temp.next.next;
            temp.next.prev = temp;

            Count--;
        }
    }
}

class program2
{
    public static void main(String arg[]) 
    {
        DoublyLLL obj= new DoublyLLL();

        obj.InsertFirst(21);
        obj.InsertFirst(11);

        obj.InsertLast(51);
        obj.InsertLast(101);

        obj.InsertAtPos(75, 4);

        obj.Display();

        System.out.println("Number of elements are : "+obj.CountNode());

        obj.DeleteAtPos(4);

        obj.DeleteFirst();

        obj.DeleteLast();

        obj.Display();

        System.out.println("Number of elements are : "+obj.CountNode());
    }
}

File: C/Data Structure/Linked List/program2.c
// Doubly linear LL

#include <stdio.h>
#include <stdlib.h>

struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

typedef struct node NODE;
typedef struct node *PNODE;
typedef struct node **PPNODE;

void InsertFirst(PPNODE Head, int No)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = No;
    newn->next = NULL;
    newn->prev = NULL;

    if (*Head == NULL)
    {
        *Head = newn;
    }
    else
    {
        newn->next = *Head;
        (*Head)->prev = newn;
        *Head = newn;
    }
}

void InsertLast(PPNODE Head, int No)
{
    PNODE newn = NULL;
    PNODE temp = NULL;

    newn = (PNODE)malloc(sizeof(NODE));

    newn->data = No;
    newn->next = NULL;
    newn->prev = NULL;

    if (*Head == NULL)
    {
        *Head = newn;
    }
    else
    {
        temp = *Head;

        while (temp->next != NULL)
        {
            temp= temp->next;
        }
        temp->next= newn;
        newn->prev= temp;
    }
}

void Display(PNODE Head)
{
    printf("Elements from linked list are: \n");

    while(Head != NULL)
    {
        printf("|%d|->", Head->data);
        Head= Head->next;
    }
    printf("NULL\n");
}

int Count(PNODE Head)
{
    int iCnt= 0;

    while(Head != NULL)
    {
        iCnt++;
        Head= Head->next;
    }
    return iCnt;
}

void InsertAtPos(PPNODE Head, int No, int Pos)
{
    int Size= 0, iCnt= 0;
    PNODE newn = NULL; 
    PNODE temp = NULL;

    Size= Count(*Head);

    if((Pos<1) || (Pos > Size +1))
    {
        printf("Invalid position \n");
        return;
    }

    if(Pos==1)
    {
        InsertFirst(Head, No);
    }

    else if(Pos==Size+1)
    {
        InsertLast(Head, No);
    }

    else
    {
        temp= *Head;

        newn= (PNODE)malloc(sizeof(NODE));

        newn->data= No;
        newn->next= NULL;
        newn->prev= NULL;

        for(iCnt=1; iCnt< Pos-1; iCnt++)
        {
            temp= temp->next;
        }
        newn->next= temp->next;
        temp->next->prev= newn;
        temp->next= newn;
        newn->prev= temp;
    }
}

void DeleteFirst(PPNODE Head)
{
    if(*Head== NULL)
    {
        return;
    }
    if((*Head)->next == NULL)
    {
        free(*Head);
        *Head= NULL;
    }
    else
    {
        *Head= (*Head)->next;
        free((*Head)->prev);
        (*Head)->prev=NULL;
    }

}

void DeleteLast(PPNODE Head)
{
    PNODE temp= NULL;

    if(*Head== NULL)
    {
        return;
    }
    else if((*Head)->next == NULL)
    {
        free(*Head);
        *Head= NULL;
    }
    else
    {
        temp= *Head;

        while(temp->next->next != NULL)
        {
            temp = temp->next;
        }
        free(temp->next);
        temp->next= NULL;
    }
}

void DeleteAtPos(PPNODE Head,int Pos)
{
    int Size= 0, iCnt= 0;

    PNODE temp= NULL;

    Size= Count(*Head);

    if((Pos<1) || (Pos > Size))
    {
        printf("Invalid position\n");
        return;
    }

    if(Pos== 1)
    {
        DeleteFirst(Head);
    }
    else if(Pos== Size)
    {
        DeleteLast(Head);
    }
    else
    {
        temp= *Head;

        for(iCnt=1; iCnt<Pos-1; iCnt++)
        {
            temp = temp->next;
        }
        temp->next = temp->next->next;
        free(temp->next->prev);
        temp->next->prev = temp;
    }
}

int main()
{
    PNODE first = NULL;
    int ret = 0;

    InsertFirst(&first,21);
    InsertFirst(&first,11);

    InsertLast(&first,51);
    InsertLast(&first,101);

    InsertAtPos(&first,75,4);

    Display(first);

    ret = Count(first);
    printf("Number of nodes are : %d\n",ret);

    DeleteAtPos(&first,4);

    DeleteFirst(&first);

    DeleteLast(&first);
    
    Display(first);

    ret = Count(first);
    printf("Number of nodes are : %d\n",ret);

    return 0;
}